.dotnet/tests/Files/FileTests.cs

```diff
@@ -22,87 +24,341 @@ public FileTests(bool isAsync) : base(isAsync)
+
+        Assert.That(visionFileInfo, Is.Null);
     }
 
     [Test]
-    public async Task UploadAndRetrieve()
+    public void ListFilesCanParseServiceError()
+    {
+        FileClient client = GetTestClient();
+        OpenAIFilePurpose fakePurpose = new OpenAIFilePurpose("world_domination");
+        ClientResultException ex = null;
+
+        if (IsAsync)
+        {
+            ex = Assert.ThrowsAsync<ClientResultException>(async () => await client.GetFilesAsync(fakePurpose));
+        }
+        else
+        {
+            ex = Assert.Throws<ClientResultException>(() => client.GetFiles(fakePurpose));
+        }
+
+        Assert.That(ex.Status, Is.EqualTo(400));
+    }
+
+    public enum FileSourceKind
+    {
+        UsingStream,
+        UsingFilePath,
+        UsingBinaryData
+    }
+
+    private static Array s_fileSourceKindSource = Enum.GetValues(typeof(FileSourceKind));
+
+    [Test]
+    [TestCaseSource(nameof(s_fileSourceKindSource))]
+    public async Task UploadFile(FileSourceKind fileSourceKind)
+    {
+        FileClient client = GetTestClient();
+        string filename = "images_dog_and_cat.png";
+        string path = Path.Combine("Assets", filename);
+        OpenAIFileInfo fileInfo = null;
+
+        try
+        {
+            if (fileSourceKind == FileSourceKind.UsingStream)
+            {
+                using Stream file = File.OpenRead(path);
+
+                fileInfo = IsAsync
+                    ? await client.UploadFileAsync(file, filename, FileUploadPurpose.Vision)
+                    : client.UploadFile(file, filename, FileUploadPurpose.Vision);
+            }
+            else if (fileSourceKind == FileSourceKind.UsingFilePath)
+            {
+                fileInfo = IsAsync
+                    ? await client.UploadFileAsync(path, FileUploadPurpose.Vision)
+                    : client.UploadFile(path, FileUploadPurpose.Vision);
+            }
+            else if (fileSourceKind == FileSourceKind.UsingBinaryData)
+            {
+                using Stream file = File.OpenRead(path);
+                BinaryData content = BinaryData.FromStream(file);
+
+                fileInfo = IsAsync
+                    ? await client.UploadFileAsync(content, filename, FileUploadPurpose.Vision)
+                    : client.UploadFile(content, filename, FileUploadPurpose.Vision);
+            }
+            else
+            {
+                Assert.Fail("Invalid source kind.");
+            }
+        }
+        finally
+        {
+            if (fileInfo != null)
+            {
+                await client.DeleteFileAsync(fileInfo.Id);
+            }
+        }
+
+        long expectedSize = new FileInfo(path).Length;
+        long unixTime2024 = (new DateTimeOffset(2024, 01, 01, 0, 0, 0, TimeSpan.Zero)).ToUnixTimeSeconds();
+        string expectedFilename = (fileSourceKind == FileSourceKind.UsingFilePath) ? path : filename;
```

Comment:
When using the path overload, the whole path is used as the filename, which creates a minor difference in behavior between overloads. Do we want to keep it like this or should we exclude the path from the filename?

Reply:
Let's keep it. It's probably better to not get in the business of parsing the file path. If users need to use a specific name, they can use the overload.